astropy/modeling/parameters.py

```diff
@@ -504,9 +504,13 @@ def shape(self):
             model_axis = self._model._model_set_axis
 
             if model_axis < 0:
-                model_axis = len(shape) + model_axis
+                shape = self.value[model_axis].shape
+                return shape
 
-            shape = shape[:model_axis] + shape[model_axis + 1:]
+            shape = shape[model_axis:]
+            shape = (shape[0] // len(self._model),) + shape[model_axis+1 :]
+            if shape[0] == 1:
+                shape = shape[1:]
 
         return shape
```

Comment:
Where was it that the use of this `shape` was previously leading evaluation with `model_set_axis=False` to fail? Was it in `_prepare_inputs_model_set`? It's not completely obvious to me that the old version of `Parameter.shape` isn't doing what was intended and that the problem isn't somewhere like the main if-else block of `_prepare_inputs_model_set` itself. But as I say, I only half follow the relatively complicated things that these thousands of lines of code are doing...

Reply:
Oh, I think I see... Maybe `pivot = input_ndim - len(max_param_shape)` at L3341? If I change that line to `pivot = input_ndim`, my failures with `model_set_axis=False` appear to go away (whatever `model_set_axis` I instantiate with). I haven't checked the correctness of that thoroughly though.